### DISCUSSION TITLE
debug时指定环境变量和工作目录？

### DISCUSSION THREAD

**Discussion by pingsoli** [xmake-io/xmake#1470] [Q&A]
### 描述讨论详情
linux下，xmake r -d target，默认会使用gdb调试target，但是target需要依赖其他库，需要指定指定工作目录。我目前的做法是使用脚本：

```shell
my_out_exe=$PWD/../bin/linux/debug/my_exe_server

cd /home/test/project/server
export LD_LIBRARY_PATH=$PWD/../third_party

my_exe_server "$@"
```

**Comment by waruqi** [comment]
set_rundir
set_runenv
add_runenvs
xmake r -w workdir
...

具体看文档

**Reply by pingsoli** [reply]
> If the user wants to modify the load directory, one is to customize the run logic by on_run(), and to do the switch inside, but just to cut the directory, this is too cumbersome.

I need to do some copying or moving tasks in my on_run(), so the `set_rundir` is not working. 

How to modify the load directory? I tried `os.cd(my_expected_dir)` in on_run(), but that does not work.

**Reply by waruqi** [reply]
你都重写 on_run 了。吧内置的 run 运行逻辑都干掉了。set_rundir 当然没效果了。。。自己 os.cd 设置，肯定可以

**Reply by pingsoli** [reply]
经测试：
```lua
...
set_rundir(my_expected_dir)
add_runenvs("LD_LIBRARY_PATH", "./xxxx/whatever/")
...
```
运行正常。
<hr>

```lua
add_runenvs("LD_LIBRARY_PATH", "./xxxx/whatever/")
on_run(function(target)
  import("core.base.option")
  os.run("echo some builtin shell commands")
  local out_exe_file=path.absolute(target:targetfile())
  os.cd(my_expected_dir)
  os.execv(out_exe_file, option.get("arguments"))
end)
```
out_exe_file读取配置文件时，路径错误，无法运行。

上面两种写法应该最后的效果是一致的吧。

**Reply by waruqi** [reply]
my_expected_dir  不要跨脚本域传递，否则取不到。我做隔离了。。另外 重写了 on_run 等于完全自己接管。。add_runenvs 就不会生效了，得自己脚本设置

**Reply by waruqi** [reply]
my_expected_dir  不要跨脚本域传递，否则取不到。我做隔离了。。另外 重写了 on_run 等于完全自己接管。。add_runenvs 就不会生效了，得自己脚本设置
